### PR TITLE
[WIP][v6r20] Space token usage

### PR DESCRIPTION
**Pull request by chaen**
This PR is meant to replace https://github.com/DIRACGrid/DIRAC/pull/3572
It contains a real implementation of the getOccupancy for the StorageElement, as well as an implementation for GFAL2_SRM2. 
Note that the version of gfal2 we use at the moment has bugs (see comments) that will be fixed in 2.15. For the time being, I wrote workarounds. 

BEGINRELEASENOTES
*Core
NEW: utility to convert file sizes between different unit

*Resources
NEW: Added StorageElement.getOccupancy() method for DIP and GFAL2_SMR2 SE types

*RSS
CHANGE: Using StorageElement.getOccupancy()

ENDRELEASENOTES

### CODE REVIEW

**Comment by chaen** [comment]
@zmathe my PR fails with
``` AccountingSystem/DB/test/Test_AccountingDB.py:169: [E1101(no-member), MakeQuery.test__queryType1] Instance of 'AccountingDB' has no '_AccountingDB__queryType' member ```
Does that ring a bell to you ? 

@fstagni  can you please check the changes I did ?

File: Resources/Storage/GFAL2_SRM2Storage.py [hunk 1]
```diff
@@ -166,35 +174,42 @@ def __getSingleTransportURL(self, path, protocols=False):
...
-      occupancy = occupancyResult['Value']
 
-    if occupancy[0] != 0:
-      return S_ERROR(occupancy)
-    output = occupancy[1][0]
+    # Gfal2 extended parameter name to query the space token occupancy
+    spaceTokenAttr = 'spacetoken.description?%s' % self.protocolParameters['SpaceToken']
+    # gfal2 can take any srm url as a base.
+    spaceTokenEndpoint = self.getURLBase()['Value']
+    print "%s %s" % (spaceTokenEndpoint, spaceTokenAttr)
```
Comment (fstagni) [inline Resources/Storage/GFAL2_SRM2Storage.py]: in a `debug` log?

File: Resources/Storage/GFAL2_SRM2Storage.py [hunk 2]
```diff
@@ -166,35 +174,42 @@ def __getSingleTransportURL(self, path, protocols=False):
...
+    try:
+      occupancyStr = self.ctx.getxattr(spaceTokenEndpoint, spaceTokenAttr)
+      try:
+        occupancyDict = json.loads(occupancyStr)
+      except ValueError:
```
Comment (fstagni) [inline Resources/Storage/GFAL2_SRM2Storage.py]: Can you prepend the comment below with a `FIXME`?

File: Resources/Storage/GFAL2_SRM2Storage.py [hunk 3]
```diff
@@ -166,35 +174,42 @@ def __getSingleTransportURL(self, path, protocols=False):
...
+        occupancyDict = json.loads(occupancyStr)[0]
+
+        # https://its.cern.ch/jira/browse/DMC-979
+        # We set totalsize to guaranteed size
+        # (it is anyway true for all the SEs I could test)
+        occupancyDict['totalsize'] = occupancyDict.get('guaranteedsize', 0)
+
+    except (gfal2.GError, ValueError) as e:
+      errStr = 'Something went wrong while checking for spacetoken occupancy.'
+      self.log.verbose(errStr, e.message)
```
Comment (fstagni) [inline Resources/Storage/GFAL2_SRM2Storage.py]: Why a `verbose` message? Why not at least a `warn`?

File: Resources/Storage/GFAL2_SRM2Storage.py [hunk 4]
```diff
@@ -166,35 +174,42 @@ def __getSingleTransportURL(self, path, protocols=False):
...
     sTokenDict = {}
-    sTokenDict['Endpoint'] = spaceTokenEndpoint
-    sTokenDict['Token'] = spaceToken
-    # Bytes to Terabytes #FIXME: have to harmonize with other SE types ---> use MB?
-    sTokenDict['Total'] = float(output.get('totalsize', '0')) / 1e12
-    sTokenDict['Guaranteed'] = float(output.get('guaranteedsize', '0')) / 1e12
-    sTokenDict['Free'] = float(output.get('unusedsize', '0')) / 1e12
+
+    sTokenDict['Total'] = float(occupancyDict.get('totalsize', '0')) / 1e6
+    sTokenDict['Free'] = float(occupancyDict.get('unusedsize', '0')) / 1e6
```
Comment (fstagni) [inline Resources/Storage/GFAL2_SRM2Storage.py]: I don't know if we use `Guaranteed` anywhere...? What was it for, do you know?

File: Resources/Storage/StorageElement.py [hunk 5]
```diff
@@ -126,7 +126,6 @@ class StorageElementItem(object):
                              "createDirectory": "createDirectory",
                              "putDirectory": "putDirectory",
                              "getDirectory": "getDirectory",
-                             "getOccupancy": "getOccupancy"
```
Comment (fstagni) [inline Resources/Storage/StorageElement.py]: Why not?

File: Resources/Storage/StorageElement.py [hunk 6]
```diff
@@ -315,6 +317,39 @@ def getStatus(self):
       return valid
     return S_OK(self.status())
 
+  def getOccupancy(self, unit='MB', **kwargs):
+    """ Retrieves the space information about the storage.
+        It returns the Total, Guaranteed and Free space .
+
+        It loops over the different Storage Plugins to query it.
+
+        :returns: S_OK with dict (keys: Total, Guaranteed, Free)
```
Comment (fstagni) [inline Resources/Storage/StorageElement.py]: Not the `Guaranteed`, it seems.

**Comment by zmathe** [comment]
__queryType is exists. I do not know why it is failing.

File: Resources/Storage/StorageElement.py [hunk 7]
```diff
@@ -641,7 +686,9 @@ def negociateProtocolWithOtherSE(self, sourceSE, protocols=None):
 
     log = self.log.getSubLogger('negociateProtocolWithOtherSE', child=True)
 
-    log.debug("Negociating protocols between %s and %s (protocols %s)" % (sourceSE.name, self.name, protocols))
+    log.debug(
+        "Negociating protocols between %s and %s (protocols %s)" %
```
Comment (fstagni) [inline Resources/Storage/StorageElement.py]: NegoTiating

**Comment by fstagni** [comment]
There's a strange failure in travis (pylint), please just merge upstream/integration.

**Comment by fstagni** [comment]
Actually integration fails: https://travis-ci.org/DIRACGrid/DIRAC/builds/330842302. I will have a look.

**Comment by fstagni** [comment]
I merge integration in this branch so that Travis should succeed.

**Comment by fstagni** [comment]
pylint in travis complains about 
```
Resources/Storage/GFAL2_SRM2Storage.py:15: [E0401(import-error), ] Unable to import 'gfal2'
```
can we add it to https://github.com/DIRACGrid/DIRAC/blob/integration/requirements.txt like we added fts3?

**Comment by chaen** [comment]
We MAY try with https://gitlab.cern.ch/dmc/gfal2-bindings
But why it does not fail with the other files ? Because I did not modify these lines ?

**Comment by fstagni** [comment]
In theory it should complain everywhere, but pylint isn't always spotting everything everywhere.

**Comment by chaen** [comment]
We could do it, but it requires quite some compilation work, so I think we just ignore it with a pylint flag. I will do it now, and rebase

**Pull request closed without merging**